server/controllers/finance/conventions.js

```diff
@@ -0,0 +1,163 @@
+
+  var query =
+      'SELECT `debitor_group`.`uuid` FROM `debitor_group` ' +
+      'WHERE `debitor_group`.`account_id` = ?';
+
+  db.exec(query, [qData.account_id])
+  .then(getConventionUuid)
+  .spread(getGroupInvoices)
+  .then(getUnpayedInvoices)
+  .then(getExchangeRate)
+  .then(handleExchangeRate)
+  .then(writePay)
+  .then(writeToJournal)
+  .catch(next)
+  .done();
+
+  function getConventionUuid(rows) {
+    /** return data to debtorGroups.fetchInvoices(id, balanced, codes) */
+    return [rows[0].uuid, null, req.codes];
+  }
+
+  function getUnpayedInvoices(rows) {
+    var situations = rows.filter(function (situation){
+      return situation.balance > 0;
+    });
+    overviews = situations;
+  }
+
+  function getExchangeRate() {
+    if (!qData.date) { throw new req.codes.ERR_DATE_NOT_DEFINED(); }
+    query =
+        'SELECT e.id, e.enterprise_id, e.currency_id, e.rate, e.date FROM exchange_rate e ' +
+        'WHERE DATE(e.date) = DATE(?) ';
+    return db.exec(query, [qData.date]);
+  }
+
+  function handleExchangeRate(rates) {
+    if (!rates.length) { throw new req.codes.ERR_EXCHANGE_RATE_NOT_FOUND(); }
+    exchangeRate = rates[0];
+  }
+
+  function writePay() {
+    /** Initialise the transaction handler */
+    var transaction = db.transaction();
+
+    /** Query to the primary cash */
+    query = 'INSERT INTO primary_cash ' +
+        '(uuid, project_id, type, date, currency_id, account_id, cost, description, cash_box_id, origin_id, user_id) ' +
+        'VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)';
+    var paramPrimaryCash = [
+        qData.uuid, qData.project_id, qData.type, new Date(qData.date), qData.currency_id,
+        qData.account_id, qData.cost, qData.description, qData.cash_box_id, qData.origin_id, qData.user_id
+    ];
+    transaction.addQuery(query, paramPrimaryCash);
+
+    /** Query to the primary cash item */
+    primaryCashItems = getPrimaryCashItems(qData.cost, qData.uuid);
+    query = 'INSERT INTO `primary_cash_item` ' +
+        '(uuid, primary_cash_uuid, debit, credit, inv_po_id, document_uuid) VALUES ?';
+    transaction.addQuery(query, [primaryCashItems]);
+
+    /** Execute queries through the transaction */
+    return transaction.execute();
+  }
+
+  function writeToJournal() {
+    journalConvention(qData.uuid, qData.user_id, function (err, result) {
+      if (err) { return next(err); }
+      res.status(201).json({ id : qData.uuid });
+    });
+  }
+
+  /** Get Primary Cash Items */
+  function getPrimaryCashItems(maxAmount, primaryCashUuid) {
+    var items = [];
+    var costReceived = maxAmount;
+
+    /** Enterprise currency is dollars */
+    var isDollars = qData.currency_id === req.session.enterprise.currency_id;
+
+    /** Sort receipts by date ASC */
+    overviews.sort(function (a, b) {
+      return a.trans_date > b.trans_date;
+    });
+
+    /** Collecting primary cash items */
+    var row, balance, allocatedCost;
+    for (var i = 0; i < overviews.length; i++) {
```

Comment:
This is a really nice algorithm.  It is very readable and the variables are clearly named.